src/varname.jl

```diff
@@ -334,37 +342,125 @@ macro vinds(expr::Union{Expr, Symbol})
 end
 
 
+@static if VERSION < v"1.5.0-DEV.666"
+    _replace_ref_begin_end(ex, withex) = Base.replace_ref_end_!(copy(ex), withex)
+    _replace_ref_begin_end(ex::Symbol, withex) = Base.replace_ref_end_!(ex, withex)
+    _index_replacement_for(s) = :($lastindex($s))
+    _index_replacement_for(s, n) = :($lastindex($s, $n))
+else
+    _replace_ref_begin_end(ex, withex) = Base.replace_ref_begin_end_!(copy(ex), withex)
+    _replace_ref_begin_end(ex::Symbol, withex) = Base.replace_ref_begin_end_!(ex, withex)
+    _index_replacement_for(s) = :($firstindex($s)), :($lastindex($s))
+    _index_replacement_for(s, n) = :($firstindex($s, $n)), :($lastindex($s, $n))
+end
+
+
 """
     vinds(expr)
 
 Return the indexing part of the [`@varname`](@ref)-compatible expression `expr` as an expression
-suitable for input of the [`VarName`](@ref) constructor.
+suitable for input of the [`VarName`](@ref) constructor (i.e., a tuple of tuples).
 
 ## Examples
 
 ```jldoctest
-julia> vinds(:(x[end]))
-:((((lastindex)(x),),))
+julia> x = [10, 20, 30]; eval(vinds(:(x[end])))
+((3,),)
+
+
+julia> x = [10 20]; eval(vinds(:(x[1, end])))
+((1, 2),)
+
+julia> x = [[1, 2]]; eval(vinds(:(x[1][end])))
+((1,), (2,))
+
+julia> x = [fill([[10], [20, 30]], 2, 2, 2)]
+       if VERSION < v"1.5.0-DEV.666"
+            eval(vinds(:(x[1][2, end, :][2][end])))
+       else
+            eval(vinds(Meta.parse("x[begin][2, end, :][2][end]")))
+       end
+((1,), (2, 2, Colon()), (2,), (2,))
 
-julia> vinds(:(x[1, end]))
-:(((1, (lastindex)(x, 2)),))
 ```
 """
-function vinds end
+function vinds(expr, head = vsym(expr))
+    # see https://github.com/JuliaLang/julia/blob/bb5b98e72a151c41471d8cc14cacb495d647fb7f/base/views.jl#L17-L75
+    indexing = _straighten_indexing(expr)
+    inds = Expr[]  # collection of result indices
+    partial = head  # partial :ref expressions, used in caching
+    cached_exprs = Vector{Pair{Symbol, Expr}}()  # cache for partial expressions in a :let
+    
+    for ixs in indexing
+        # S becomes the name of the cached variable
+        S = (partial == head) ? head : gensym(:S)
+        used_S = false
+        
+        nixs = length(ixs)
+        if nixs == 1
+            # for 1D indexing, just use `lastindex(x)`
```

Comment:
```suggestion
            # For linear indexing we use `lastindex(x)`.
```
I was a bit confused as to even if we're doing 1D-indexing, why don't just use use `_index_replacement_for(S, 1)`, but then I remembered there's such a thing as linear indexing:) So I think this is a bit clearer.